### PR TITLE
Add utility to test if autograd was registered correctly

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106558
* __->__ #106561

See docstring for more details. This API is not meant to be directly
user-facing, it is meant to be used as a subtest of D47965247, which is
coming soon.

Test Plan:
- Some new tests.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/testing/_internal/optests/autograd_registration.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    requires_grad=True so that the autograd path gets exercised,
+    otherwise, this check may erroneously pass.
+
+    Here are some best practices if you do find your autograd is
+    registered incorrectly:
+    - If the operator is composite (i.e. consists of other PyTorch ops)
+      and you wish the operator to decompose and get autograd support
+      that way, then please register the implementation to
+      DispatchKey::CompositeImplicitAutograd
+    - If you're adding an autograd formula for the operator, the correct
```
Comment (soulitzer) [inline torch/testing/_internal/optests/autograd_registration.py:40]: Should we recommend the custom op API instead/too?

File: torch/testing/_internal/optests/autograd_registration.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    #   CompositeImplicitAutograd or not an op) or if the user invokes
+    #   an autograd.Function in the backend key.
+    #
+    # Note that it's already likely a bug if the operator directly returns
+    # an input as output (because custom ops don't have a good way of
+    # constructing true in-place or out variants), but we defer that
+    # responsibility to a different test (schema_check).
+
+    # If no inputs require grad, we are unable to perform this test, so we
+    # just succeed.
```
Comment (soulitzer) [inline torch/testing/_internal/optests/autograd_registration.py:78]: Maybe we should error instead?

File: torch/testing/_internal/optests/autograd_registration.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    # Our strategy for detecting if an operator doesn't have an autograd
+    # kernel at the autograd key is:
+    # - set the autograd fallback mode to "nothing" (so it does not change
+    #   the required-gradness of outputs)
+    # - run the operator
+    # - Check if any outputs of the operator (that are not inputs) require
+    #   grad. This would only happen if the user calls regular PyTorch
+    #   operations in their backend key (this op should instead be
+    #   CompositeImplicitAutograd or not an op) or if the user invokes
+    #   an autograd.Function in the backend key.
```
Comment (soulitzer) [inline torch/testing/_internal/optests/autograd_registration.py:70]: What if the op is partially a composite op and partially does non-pytorch operations? Are we sure we also want to error in that case? I guess in theory this means they could extract out a smaller custom operation. I'm not sure how often this happens in practice though.

If we wanted to test that the operator is "fully" composite that would probably mean having to actually validate the gradients are wrong, but maybe that is too complex to be worth?

File: torch/testing/_internal/optests/autograd_registration.py (hunk 1 above)
Comment (zou3519) [reply torch/testing/_internal/optests/autograd_registration.py:40]: In the future, when the custom op guide is done, I am going to link it from everywhere. I haven't written about the autograd API yet so it wasn't mentioned here

File: torch/testing/_internal/optests/autograd_registration.py (hunk 3 above)
Comment (zou3519) [reply torch/testing/_internal/optests/autograd_registration.py:70]: > What if the op is partially a composite op and partially does non-pytorch operations? Are we sure we also want to error in that case?

Yup. Ideally the user makes the custom op as small as possible.

> If we wanted to test that the operator is "fully" composite that would probably mean having to actually validate the gradients are wrong, but maybe that is too complex to be worth?

Yeah, this is meant to be a simple test.

File: torch/testing/_internal/optests/autograd_registration.py (hunk 2 above)
Comment (zou3519) [reply torch/testing/_internal/optests/autograd_registration.py:78]: It works a bit better for the crossref testing if this doesn't actually error, but if we feel strongly about the error I can make it error with a special error and then surpress it during crossref testing.

Comment (soulitzer) [reply torch/testing/_internal/optests/autograd_registration.py:78]: My vote would be for erroring and then suppressing where necessary

Comment (soulitzer) [reply torch/testing/_internal/optests/autograd_registration.py:78]: Oh I just checked, gradcheck actually errors if no inputs require grad! so probably good to do the same here.

Comment (zou3519) [reply torch/testing/_internal/optests/autograd_registration.py:78]: Sounds good!

File: torch/testing/_internal/optests/autograd_registration.py (hunk 3 above)
Comment (soulitzer) [reply torch/testing/_internal/optests/autograd_registration.py:70]: Ah makes sense. I don't know how common it is for folks to mix normal torch ops in with their custom ops, but if it is, it might be worth mentioning this case in the error/docs.

**Review by soulitzer — APPROVED** [review]
